manager/container.go

```diff
@@ -572,9 +579,116 @@ func (c *containerData) updateStats() error {
+		}
+		powerUsage, err := device.PowerUsage()
+		if err != nil {
+			return fmt.Errorf("error while getting gpu power usage: %v", err)
+		}
+
+		stats.NvidiaGpu = append(stats.NvidiaGpu, info.NvidiaGpuStats{
+			MinorNumber: uint64(minor),
+			Name:        name,
+			MemoryTotal: memoryTotal,
+			MemoryUsed:  memoryUsed,
+			Busy:        uint64(utilizationGPU),
+			PowerUsage:  uint64(powerUsage),
+		})
+	}
+	return nil
+}
+
+// getNvidiaMinorNumbers returns a list of minor numbers corresponding to NVIDIA GPU
+// devices that the container is allowed to access according to its devices cgroup
+func (c *containerData) getNvidiaMinorNumbers() []int {
+	if c.nvidiaMinorNumbers == nil {
+		var err error
+		c.nvidiaMinorNumbers, err = c.parseDevicesCgroup()
+		if err != nil {
+			glog.Infof("GPU metrics may be unavailable/incomplete for container %q: failed to parse devices cgroup: %v", c.info.Name, err)
+		}
+	}
+	return c.nvidiaMinorNumbers
+}
+
+// parseDevicesCgroup parses the devices cgroup devices.list file for the container
+// and returns a list of minor numbers corresponding to NVIDIA GPU devices that the
+// container is allowed to access. In cases where the container has access to all
+// devices or all NVIDIA devices but the devices are not enumerated separately in
+// the devices.list file, we return an empty list.
+func (c *containerData) parseDevicesCgroup() ([]int, error) {
+	// Always return a non-nil slice
+	nvidiaMinorNumbers := []int{}
+
+	devicesCgroupPath, err := c.handler.GetCgroupPath("devices")
+	if err != nil {
+		return nvidiaMinorNumbers, fmt.Errorf("error getting devices cgroup path: %v", err)
+	}
+
+	devicesList := devicesCgroupPath + "/devices.list"
+	f, err := os.Open(devicesList)
+	if err != nil {
+		return nvidiaMinorNumbers, fmt.Errorf("error while opening devices cgroup file %q: %v", devicesList, err)
+	}
+	defer f.Close()
+
+	s := bufio.NewScanner(f)
+
+	// See https://www.kernel.org/doc/Documentation/cgroup-v1/devices.txt for the file format
+	for s.Scan() {
+		text := s.Text()
+
+		fields := strings.Fields(text)
+		if len(fields) != 3 {
+			return nvidiaMinorNumbers, fmt.Errorf("invalid devices cgroup entry %q: must contain three whitespace-separated fields", text)
+		}
+
+		// Split the second field to find out major:minor numbers
+		majorMinor := strings.Split(fields[1], ":")
+		if len(majorMinor) != 2 {
+			return nvidiaMinorNumbers, fmt.Errorf("invalid devices cgroup entry %q: second field should have one colon", text)
+		}
+
+		// NVIDIA graphics devices are character devices with major number 195.
+		// https://github.com/torvalds/linux/blob/v4.13/Documentation/admin-guide/devices.txt#L2583
+		if fields[0] == "c" && majorMinor[0] == "195" {
+			minorNumber, err := strconv.Atoi(majorMinor[1])
+			if err == nil {
+				nvidiaMinorNumbers = append(nvidiaMinorNumbers, minorNumber)
+			}
+			// We are ignoring the "195:*" case
+			// where the container has access to all NVIDIA devices on the machine.
```

Comment:
This comment slightly confused me, a container can have access to all GPUs on the machine but not have `195:*`, for instance if it enumerates each device:
```
c 195:3 rwm
c 195:2 rwm
c 195:1 rwm
c 195:0 rwm
```
It depends on what the container runtime is doing.

What is the rationale behind skipping this use case?